# Mini projeto - Técnicas de Programação I

**Aluno:** Ítalo da Silva Monteiro

**Turma:** Unimed Campinas Data Academy | #1108

## Objetivo
Criar uma simulação para um jogo de dados. Essa simulação tem como objetivo reunir estatísticas para analisar a justiça do jogo, possíveis resultados e fazer previsões sobre jogos futuros.

## Desafios do Projeto
1- Simulação de Dados: Crie uma função que simule o lançamento de dois dados de seis lados (valores de 1 a 6). Esta função deve retornar a soma dos resultados dos dados.

2- Múltiplas Simulações: Use a função do passo 1 para simular um grande número de jogos de dados (digamos, 1000 jogos). Armazene o resultado de cada jogo em um array NumPy.

3- Análise de Dados: Agora, vamos analisar os resultados desses jogos. Calcule e imprima o seguinte:

A média dos resultados.  
O lançamento máximo e mínimo.  
O número de vezes que cada possível lançamento (2, 3, 4, 5, 6, 7, 8, 9, 10, 11 e 12) ocorreu.  

Teste de Hipótese: Agora vamos fazer um pouco de teste de hipóteses:  
Supondo um jogo justo (ou seja, todos os lançamentos são igualmente prováveis), o resultado da sua simulação coincide com essa suposição?  
Por que sim ou por que não?  
O que isso significa para um jogador do jogo de dados?

In [15]:
import numpy as np

In [16]:
def simulacao_dados():
    """
    A função simula o lançamento de dois dados de seis lados e retorna a soma dos resultados.

    Esta função não recebe nenhum parâmetro. Ela usa a biblioteca numpy para gerar dois
    números aleatórios entre 1 e 6, representando os resultados de lançar dois dados.

    Retorna:
        int: A soma dos dois números gerados, simulando a soma dos resultados de dois dados.

    Exemplo:
        --> lancamento_dados_numpy()
        7  # O resultado pode variar a cada chamada, pois é aleatório.
    """
    primeiro_dado = np.random.randint(1,7)
    segundo_dado = np.random.randint(1,7)
    
    return primeiro_dado + segundo_dado

print(f'O resultado da simulação foi: {simulacao_dados()}.')

O resultado da simulação foi: 9.


In [17]:
np.random.seed(42) # definindo a semente para que os resultados aleatórios sejam fixos.
num_jogos = np.array([simulacao_dados() for _ in range(1000)]) # Chama a função "simulacao_dados" 1000 vezes. 
num_jogos.sort() # Ordena o array em ordem crescente
print(f'Array criado --> {num_jogos}') # Mostra na tela o conteúdo do array ordenado.
print(f'Soma --> {num_jogos.sum()}') # Soma todos os resultados dos jogos e mostra na tela.

Array criado --> [ 2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2
  2  2  2  2  2  2  2  2  2  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3  3
  3  3  3  3  3  3  3  3  3  3  3  3  4  4  4  4  4  4  4  4  4  4  4  4
  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4
  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4
  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  4  5  5  5  5  5
  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5
  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5
  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5
  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5
  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  5  6  6  6  6  6  6  6
  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6  6
  6  6  6  6  6  6  6  6  6  6  6 

In [18]:
media = round(num_jogos.mean(),2) # Calcula a média da soma de todos os jogos e mostra com 2 casas após o ponto.
lancamento_max = num_jogos.max() # Traz o maior resultado da soma dos dois lados nas 1000 rodadas.
lancamento_min = num_jogos.min() # Traz o menor resultado da soma dos dois lados nas 1000 rodadas.
valores, contagens = np.unique(num_jogos, return_counts=True) # Traz os valores únicos e a quantidade de vezes que ele apareceu
frequencias = dict(zip(valores, contagens)) # cria um dicionário onde a chave será o array "valores" (únicos) e o valor será o array "contagens" (vezes que apareceu)

print(f'A média dos resultados é: {media}\n'
      f'O lançamento máximo foi: {lancamento_max}\n'
      f'O lançamento mínimo foi: {lancamento_min}\n'
      f'Veja a frequência dos resultados: {frequencias}')

A média dos resultados é: 6.96
O lançamento máximo foi: 12
O lançamento mínimo foi: 2
Veja a frequência dos resultados: {2: 33, 3: 51, 4: 79, 5: 118, 6: 166, 7: 136, 8: 145, 9: 111, 10: 72, 11: 63, 12: 26}


**Teste de hipótese:**

Em um jogo justo de dois dados de seis lados, é esperado que os resultados sejam igualmente prováveis. Mas, a soma de dois lados não tem a mesma probabilidade. Exemplo, a soma **7** tem mais possibilidades **(6/36)**, pois há mais combinações **(1+6, 2+5, 3+4, 4+3, 5+2, 6+1)** que resultam em **7**. Já, as somas **2** e **12** são as menos prováveis **(1/36 para cada)**, pois há apenas uma combinação para cada número **(1+1 para 2 e 6+6 para 12)**.

No meu caso os resultados se alinham com a expectativa teórica. Os números centrais (especialmente em torno de 6, 7 e 8) têm frequências mais altas, o que é esperado, pois existem mais combinações que resultam nesses números.
Por exemplo, o número **7**, que teoricamente tem a maior probabilidade, ocorreu **136** vezes, que é uma alta frequência. O número **6** ocorreu **166** vezes, o que é interessante, pois indica uma variação que pode ocorrer em simulações.

Para um jogador, isso reforça a ideia de que, embora certos resultados sejam mais prováveis, o acaso desempenha um papel significativo em qualquer jogo individual, e é bom estarmos cientes de que, mesmo que alguns resultados sejam teoricamente mais prováveis, isso não garante que eles ocorrerão mais frequentemente em uma série curta de jogos.